activesupport/test/core_ext/time_ext_test.rb

```diff
@@ -887,13 +887,16 @@ def test_at_with_datetime_returns_local_time
 
   def test_at_with_time_with_zone
     assert_equal Time.utc(2000, 1, 1, 0, 0, 0), Time.at(ActiveSupport::TimeWithZone.new(Time.utc(2000, 1, 1, 0, 0, 0), ActiveSupport::TimeZone["UTC"]))
+    assert_equal Time.utc(2000, 1, 1, 0, 0, 0), Time.at(ActiveSupport::TimeWithZone.new(Time.utc(2000, 1, 1, 0, 0, 0), ActiveSupport::TimeZone["UTC"]), 0)
+  end
 
-    # Only test this if the underlying Time.at raises a TypeError
-    begin
-      Time.at_without_coercion(Time.now, 0)
-    rescue TypeError
-      assert_raise(TypeError) { assert_equal(Time.utc(2000, 1, 1, 0, 0, 0), Time.at(ActiveSupport::TimeWithZone.new(Time.utc(2000, 1, 1, 0, 0, 0), ActiveSupport::TimeZone["UTC"]), 0)) }
-    end
+  def test_at_with_time_precision
+    time_with_nsec = Time.at(500000000, 123456789, :nsec)
+    time_with_zone = ActiveSupport::TimeWithZone.new(time_with_nsec, ActiveSupport::TimeZone["UTC"])
+    assert_equal time_with_zone, Time.at(time_with_zone)
+    assert_equal time_with_zone.to_r, Time.at(time_with_zone).to_r
+    assert_equal time_with_zone.to_f, Time.at(time_with_zone).to_f
+    assert_equal time_with_zone.nsec, Time.at(time_with_zone).nsec
```

Comment:
Are we missing an equivalent test for DateTime? Its precision is also improved by these changes.